search-endpoint.md

```diff
@@ -81,7 +78,6 @@ The JSON response document is as follow
 
 Exposed links:
 * objects: link to get the actual list of objects that match the search query
-* facets: link to get the list of facet values and counts associated with this search query as configured in https://github.com/DSpace/DSpace/blob/master/dspace/config/spring/api/discovery.xml#L86
```

Comment:
No, the facet values associated with the search are embedded. It is not possible to give a single link that returns values for all configured facets.

Reply:
There is something conceptually wrong here. You cannot embed something that is not addressable. In addition, the current RestController should provide navigation to linked resources automatically, of course the code is not optmized as it essentially retrieve the "parent" object and extract the embedded resource, see
https://github.com/DSpace/DSpace/blob/master/dspace-spring-rest/src/main/java/org/dspace/app/rest/RestResourceController.java#L186

Reply:
We can add the link. The results would then look like this:

```
{
  "query":"my query",
  "scope":"9076bd16-e69a-48d6-9e41-0238cb40d863",
  "appliedFilters": [
      {
        "filter" : "title",
        "operator" : "notcontains",
        "value" : "abcd",
        "label" : "abcd"
      },
      {
        "filter" : "author",
        "operator" : "authority",
        "value" : "1234",
        "label" : "Smith, Donald"
      }
  ],
  "sort" : {
    "by" : "dc.date.issued",
    "order" : "asc"
  },
  "page": {
    	"size": 5,
    	"totalElements": 14,
    	"totalPages": 3,
    	"number": 0
  },
  "_embedded" : {
    "facets" : [
      {
        "name" : "author",
        "facetType": "text",
        "hasMore": true,
        "facetLimit": 5,
        "_links": {
          "next": "/api/discover/facets/author?page=1&size=5",
          "self": "/api/discover/facets/author"
        },
        "page": {
          "number": 0,
          "size": 5
        },
        "_embedded" : {
          "values" : [
              {
                "value" : "Smith, Donald 2",
                "count" : 100,
                "_links": {
                  "search" : "/api/discover/search/objects?query=my+query&scope=9076bd16-e69a-48d6-9e41-0238cb40d863&f.title=abcd,notcontains&f.author=1234,authority&f.author=Smith,+Donald+2,equals"
                }
              },
              {
                "value" : "Smith, Donald 1",
                "count" : 80,
                "_links": {
                  "search" : "/api/discover/search/objects?query=my+query&scope=9076bd16-e69a-48d6-9e41-0238cb40d863&f.title=abcd,notcontains&f.author=1234,authority&f.author=Smith,+Donald+1,equals"
                }
              },
              {
                "value" : "Smith, Donald 3",
                "count" : 10,
                "_links": {
                  "search" : "/api/discover/search/objects?query=my+query&scope=9076bd16-e69a-48d6-9e41-0238cb40d863&f.title=abcd,notcontains&f.author=1234,authority&f.author=Smith,+Donald+3,equals"
                }
              }
          ]
        },
        "_links" : {
          "self": {
            "href": "/api/discover/facets/author?query=my+query&scope=9076bd16-e69a-48d6-9e41-0238cb40d863&f.title=abcd,notcontains&f.author=1234,authority&page=0&size=5"
          }
        }
      },
      {
        "name" : "subject",
        "facetType": "text",
        "hasMore": true,
        "facetLimit": 5,
        "_links": {
          "next": "/api/discover/facets/subject?page=1&size=5",
          "self": "/api/discover/facets/subject"
        },
        "page": {
          "number": 0,
          "size": 5
        },
        "_embedded" : {
          "values" : [
              {
                "value" : "Java",
                "count" : 100,
                "_links": {
                  "search" : "/api/discover/search/objects?query=my+query&scope=9076bd16-e69a-48d6-9e41-0238cb40d863&f.title=abcd,notcontains&f.author=1234,authority&subject.equals=Java"
                }
              },
              {
                "value" : "SQL",
                "count" : 80,
                "_links": {
                  "search" : "/api/discover/search/objects?query=my+query&scope=9076bd16-e69a-48d6-9e41-0238cb40d863&f.title=abcd,notcontains&f.author=1234,authority&subject.equals=Java"
                }
              },
              {
                "value" : "CSS",
                "count" : 10,
                "_links": {
                  "search" : "/api/discover/search/objects?query=my+query&scope=9076bd16-e69a-48d6-9e41-0238cb40d863&f.title=abcd,notcontains&f.author=1234,authority&subject.equals=Java"
                }
              }
          ]
        },
        "_links" : {
          "self": {
            "href": "/api/discover/facets/subject?query=my+query&scope=9076bd16-e69a-48d6-9e41-0238cb40d863&f.title=abcd,notcontains&f.author=1234,authority&page=0&size=5"
          }
        }
      }
    ]
  },
  "_links": {
      "self": {
        "href": "/api/discover/search/facets?query=my+query&scope=9076bd16-e69a-48d6-9e41-0238cb40d863&f.title=abcd,notcontains&f.author=1234,authority&page=0&size=5"
      }
  }
}
```

But since these results are already embedded, I don't think this endpoint will be used by the current UI. Is that a problem?

Reply:
It is expected, nevertheless the endpoint needs to work as other client could use it in future